Extend argmax to support inputs with tile layout

iwroszTT opened the pull request:
### Ticket
https://github.com/tenstorrent/tt-metal/issues/23005

### Problem description
Argmax operation was missing support for tiled tensor layout.

### What's changed
Tiled tensor support is added by adding a dedicated device kernel, see the reader_argmax_tile_layout.cpp file.
Tile layout is available only with single core. For multicore version, i.e. when use_muticore=True, only row major layout is available. The case of reduce_all (when dim parameter is not given) is not implemented. Only fp32 and bfloat16 data types are supported.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [x] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).

bbradelTT reviewed (commented):
Please add a unit test as well in tests/ttnn/unit_tests/operations/reduce/test_argmax.py

iwroszTT committed:
Extend argmax to support inputs with tile layout

iwroszTT committed:
Argmax tiled kernel: compute integer division in compile time

 - Argmax trace sweep: for bf16 data type, use same precision
   in ttnn and torch input

bbradelTT commented on ttnn/cpp/ttnn/operations/reduction/argmax/device/argmax_program_factory.cpp:
```diff
@@ -86,81 +86,179 @@ static inline std::tuple<CoreRangeSet, CoreRangeSet, CoreRangeSet, uint32_t, uin
+        TT_FATAL(reduce_all == false, "reduce_all not supported with TILE_LAYOUT");
+
+        // With TILE layout, padded shape has always at least 2 dims (i.e., also for 1D input tensors).
+        TT_FATAL(rank > 1, "Invalid rank for tiled tensor");
```
It would be good to move these fatals into the validation function.

bbradelTT commented on ttnn/cpp/ttnn/operations/reduction/argmax/device/kernels/argmax_tile_layout.hpp:
```diff
@@ -0,0 +1,320 @@
+
+            if (!(row_index < max_size)) {
+                WAYPOINT("AST1");
+                ASSERT(false);
```
If this never happens, you can just
`ASSERT(row_index < max_size);`
and remove the rest.

bbradelTT commented on ttnn/cpp/ttnn/operations/reduction/argmax/device/kernels/argmax_tile_layout.hpp:
```diff
@@ -0,0 +1,320 @@
+/**
+ * @brief Struct container that gathers parameters related to the output tensor
+ *
+ * Since the output tensor parameters
```
This entire comment seems to be completely off.

bbradelTT commented on ttnn/cpp/ttnn/operations/reduction/argmax/device/kernels/reader_argmax_tile_layout.cpp:
```diff
@@ -0,0 +1,214 @@
+    constexpr uint32_t output_page_elements = keepdim ? 1 : logical_height;
+    // Number of data elements to accumulate before writing data to the output tensor.
+    constexpr uint32_t stage_size = keepdim ? tile_height : logical_height;
+    uint32_t accumulated_arg_max[stage_size] = {0};
```
How large would logical_height need to be before this fails because the program on device will become too large?

Should a CB be used instead?

iwroszTT committed:
Argmax: fixes due to code review

 - Improved argmax docstring
 - Improved comments, logic related to watcher asserts
 - Small other fixes, e.g use constants from metal header file

iwroszTT committed:
Argmax: remove a redundant c_str() from program factory code

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>
